pkg/ui/server.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ui
+
+import (
+	"context"
+	"html/template"
+	"log"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// Pair represents a key value pair used by the select HTML element.
+type Pair struct {
+	Key  string
+	Text string
+}
+
+// Content defines the displayable parts of the token retrieval form.
+type Content struct {
+	UserLabel     string
+	CategoryLabel string
+	ReasonLabel   string
+	TTLLabel      string
+	Categories    []Pair
+	TTLs          []Pair
+}
+
+// FormDetails represents all the input and content used for the token retrievlal form.
+type FormDetails struct {
+	WindowName string
+	Origin     string
+	PageTitle  string
+	Content    Content
+	Category   string
+	Reason     string
+	TTL        string
+	Errors     map[string]string
+}
+
+// SuccessDetails represents the data used for the success page and the postMessage response to the client.
+type SuccessDetails struct {
+	Token        string
+	TargetOrigin string
+	WindowName   string
+}
+
+var (
+	categories []string
+	ttls       []string
+)
+
+// RunServer initializes a server on port 9091 and registers a handler for the /popup route.
+func RunServer(ctx context.Context) {
+	categories = []string{"explanation", "breakglass"}
+	ttls = []string{"15", "30", "60", "120", "240"}
+
+	router := http.NewServeMux()
+	server := &http.Server{
+		Addr:         ":9091",
```

Comment:
This needs to be configurable, likely via an envvar. We use https://github.com/sethvargo/go-envconfig